File: PwaUtils/KinUtils.cc
//************************************************************************//
//                                                                        //
//  Copyright 2016 Bertram Kopf (bertram@ep1.rub.de)                      //
//                 - Ruhr-Universität Bochum                              //
//                                                                        //
//  This file is part of Pawian.                                          //
//                                                                        //
//  Pawian is free software: you can redistribute it and/or modify        //
//  it under the terms of the GNU General Public License as published by  //
//  the Free Software Foundation, either version 3 of the License, or     //
//  (at your option) any later version.                                   //
//                                                                        //
//  Pawian is distributed in the hope that it will be useful,             //
//  but WITHOUT ANY WARRANTY; without even the implied warranty of        //
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         //
//  GNU General Public License for more details.                          //
//                                                                        //
//  You should have received a copy of the GNU General Public License     //
//  along with Pawian.  If not, see <http://www.gnu.org/licenses/>.       //
//                                                                        //
//************************************************************************//

// DataUtils class definition file. -*- C++ -*-
// Copyright 2016 Bertram Kopf

#include "PwaUtils/KinUtils.hh"
#include "PspGen/EvtVector4R.hh"
#include "PspGen/EvtKine.hh"
#include "Utils/PawianConstants.hh"

void KinUtils::SetTheta(Vector4<double>& the4Vec, double theta){
  double energy = the4Vec.E();
  double mom = the4Vec.P();
  double phi   = the4Vec.Phi();
  
  double px = mom*sin(theta)*cos(phi);
  double py = mom*sin(theta)*sin(phi);
  double pz = mom*cos(theta);
  
  the4Vec.SetV4(energy, px, py, pz);
}

void KinUtils::SetPhi(Vector4<double>& the4Vec, double phi) 
{
  double p_xy   = the4Vec.Pxy();
  
  double energy = the4Vec.E();
  double px = p_xy*cos(phi);
  double py = p_xy*sin(phi);      
  double pz = the4Vec.Pz();
  
  the4Vec.SetV4(energy, px, py, pz);
}

Vector4<double>  KinUtils::heliEvtGenVec(Vector4<double>& motherRef, Vector4<double>& ref, Vector4<double>& mother, 
				   Vector4<double>& daughter) {
  //use EvtGen classes for calculating theta and phi; temporary solution and should be fixed soon
  EvtVector4R motherRefEvtGen(motherRef.E(), motherRef.Px(), motherRef.Py(), motherRef.Pz());  
  EvtVector4R refEvtGen(ref.E(), ref.Px(), ref.Py(), ref.Pz());
  EvtVector4R motherEvtGen(mother.E(), mother.Px(), mother.Py(), mother.Pz());
  EvtVector4R daughterEvtGen(daughter.E(), daughter.Px(), daughter.Py(), daughter.Pz()); 
  
  double thetaHeliDaughter= acos(EvtDecayAngle(refEvtGen, motherEvtGen, daughterEvtGen));
  double phiHeliDaughter= EvtDecayAnglePhi(motherRefEvtGen, refEvtGen, motherEvtGen, daughterEvtGen);
  
  Vector4<double> result(daughter);
  result.Boost(mother);
  KinUtils::SetTheta(result, thetaHeliDaughter);
  KinUtils::SetPhi(result, phiHeliDaughter);
  
  return result;
}

template<typename T> 
Vector4<T> KinUtils::heliVec(const Vector4<T>& motherRef,  const Vector4<T>& ref,const Vector4<T>& mother, const Vector4<T>& daughter){
  Vector4<T> result=daughter;
  Vector4<T> refTrafo=ref;
  Vector4<T> refRecoilTrafo=motherRef-ref;

  //boost all vectors into the mother rest frame
  result.Boost(mother);
  refTrafo.Boost(mother);
  refRecoilTrafo.Boost(mother);

  //rotate vectors so that refRecoilTrafo moves in the negative direction of the z-axis   
  result.RotateZ(-refTrafo.Phi());
  result.RotateY(PawianConstants::pi-refTrafo.Theta());

  refRecoilTrafo.RotateZ(-refTrafo.Phi());
  refRecoilTrafo.RotateY(PawianConstants::pi-refTrafo.Theta());

  //rotate around the z-axis so that refRecoil lies in the x-z plain
  result.RotateZ(PawianConstants::pi-refRecoilTrafo.Phi()); 
  
  return result;
}

template<typename T>
Vector4<T> KinUtils::perpTo(const Vector4<T>& vecA,  const Vector4<T>& vecB){

  Vector4<T> result(vecA.E()+vecB.E(),
		    vecA.Y()*vecB.Z()-vecA.Z()*vecB.Y(),
		    vecA.Z()*vecB.X()-vecA.X()*vecB.Z(),
		    vecA.X()*vecB.Y()-vecA.Y()*vecB.X());
  // energy is arbitrary; use sum of vector A and vector B here
 return result;
} 


template Vector4<double> KinUtils::heliVec(const Vector4<double>&,  const Vector4<double>&,const Vector4<double>& , const Vector4<double>& );  
template Vector4<float> KinUtils::heliVec(const Vector4<float>&,  const Vector4<float>&,const Vector4<float>& , const Vector4<float>& );
template Vector4<double> KinUtils::perpTo(const Vector4<double>& vecA,  const Vector4<double>& vecB);
template Vector4<float> KinUtils::perpTo(const Vector4<float>& vecA,  const Vector4<float>& vecB);

File: Scripts/Mathematica/CMs/KinUtils.m
(*  Copyright 2025                                                                                
 Bertram Kopf (bertram@ep1.rub.de)
    Ruhr-Universität Bochum

    This file is part of Pawian.

    Pawian is free software: you can redistribute it and/or modify    
    it under the terms of the GNU General Public License as published by 
    the Free Software Foundation, either version 3 of the License, or     
    (at your option) any later version.                       
                                     
    Pawian is distributed in the hope that it will be useful,         
    but WITHOUT ANY WARRANTY; without even the implied warranty of    
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the     
    GNU General Public License for more details.                  
                                     
    You should have received a copy of the GNU General Public License     
    along with Pawian.  If not, see <http://www.gnu.org/licenses/>

***********************************************************************************************
  This Mathematica-script is a colletion of functions which are needed to perform calculations with the K-matrix formalism.
 *)

(* phase space factor *)
rho[s_,m1_,m2_]:= Sqrt[1-(m1+m2)^2/s] Sqrt[1-(m1-m2)^2/s];

(* break up momentum *)
BreakupMomQDefaultFromS[s_, s1_, s2_]:=
  Module[{result},
	 result=rho[s, Sqrt[s1], Sqrt[s2]] Sqrt[s]/2.;
	 (* result = Sqrt[s*s + s1*s1 + s2*s2 - 2. * ( s*s1 + s*s2 + s1*s2)]/Sqrt[4.*s]; *)
	 result
	 ];

(* CM Function based on Dudek et al  Phys. Rev. D91, 054008 (2015), Appendix B; subtraction constant at s=s_thresh CM(s_thresh)=0 *)
xi[s_,m1_,m2_]:=1-(m1+m2)^2/s;
cmDudekPre[s_,m_,u_] := rho[s,m,u]/\[Pi] Log[(xi[s,m,u]+rho[s,m,u])/(xi[s,m,u]-rho[s,m,u])]-xi[s,m,u]/\[Pi] (u-m)/(u+m) Log[u/m];

(* making sure to be in the 1 quadrant for s=real *)
cmDudek[s_,m_,u_] := cmDudekPre[s + I * If[Im[s] == 0, 10.^-8, 0.], m, u];


(* CM Function based on Reid with subtraction constant at s=0 CM(0)=0 *)
(* cmReid[s_,m_,u_]:= - ( cmDudek[s,m,u] - 1/\[Pi] (1. + 2 m u / (u u - m m) Log[u/m] ) ); *)

(* proteted case for u==m with L'Hospital *) 
cmReid[s_,m_,u_]:= - ( cmDudek[s,m,u] - If[m != u, 1/\[Pi] (1. + 2 m u / (u u - m m) Log[u/m]), 2./\[Pi]] );

(* cmReid[s_,m_,u_]:= - ( cmDudek[s,m,u] - If[m != u, 1/\[Pi] (1. + 2 m u / (u u - m m) Log[u/m] ), 2./\[Pi]  *)


(* Blatt-Weisskopf factors *)
BlattWeisskopf[l_, z_]:=
  Module[{result},
	 result = Which[l==0, 1.0 + 0. I, l==1, Sqrt[2. z / (z+1.)] + 0. I, l==2, Sqrt[13. z z / (z z + 3. z + 9.)]+ 0. I];
	 result
	 ];

BWBarrier[l_, q_, q0_]:=
  Module[{result},
	 qR:=0.1973;
  z:=(Norm[q] Norm[q])/(Norm[qR] Norm[qR]);
	 (* Print["z: ", z]; *) 
	 z0:=(Norm[q0] Norm[q0])/(Norm[qR] Norm[qR]);
	 (* Print["z0: ", z0]; *)
	 result:= BlattWeisskopf[l,z]/BlattWeisskopf[l,z0];
	 result
	 ];

(* elasticity *)
Sii[s_, m1_, m2_, Tii_] := 1. + 2. I Sqrt[Re[rho[s,m1,m2]]] Tii Sqrt[Re[rho[s,m1,m2]]];


(* extract delta phase of an Argand plot *)
deltaArgand[mass_, m1_, m2_, Tii_]:=
  Module[{delta},
	 Tiirho :=rho[mass*mass,m1,m2] Tii;
  ReE := Re[Tiirho];
  ImE := Im[Tiirho] - 0.5;
	 tmpdelta := ArcTan[Abs[ReE], ImE];
	 delta := 0.5 tmpdelta 180./Pi + 45.0;
	 Which[ReE  < 0.0, delta = 180.0 - delta];
	 delta
	 ];

(* Breit-Wigner function w/ Blatt-Weisskopf factors *)
BreitWignerBlattWRel[sCurrent_, orbMom_, M0_, Width_, Ma_, Mb_]:=
  Module[{BW},
	 s0=M0 M0;
	 rho0=rho[s0, Ma, Mb];
	 rhoCurrent=rho[sCurrent, Ma, Mb];
         momQ0=BreakupMomQDefaultFromS[s0, Ma Ma, Mb Mb];
	 momQCurrent=BreakupMomQDefaultFromS[sCurrent, Ma Ma, Mb Mb];
	 bwBarrierFac=If[orbMom==0, 1., BWBarrier[orbMom, momQCurrent, momQ0]];
	 BW = M0 Width  bwBarrierFac / (s0 - sCurrent - I (rhoCurrent/rho0) M0 Width bwBarrierFac bwBarrierFac); 
	 BW
	 ];

BreitWignerBlattWRelNorm[sCurrent_, orbMom_, M0_, Width_, Ma_, Mb_]:=
  Module[{BW},
         s0=M0 M0;
         rho0=rho[s0, Ma, Mb];
         rhoCurrent=rho[sCurrent, Ma, Mb];
         momQ0=BreakupMomQDefaultFromS[s0, Ma Ma, Mb Mb];
         momQCurrent=BreakupMomQDefaultFromS[sCurrent, Ma Ma, Mb Mb];
         bwBarrierFac=If[orbMom==0, 1., BWBarrier[orbMom, momQCurrent, momQ0]];
         GammaM= Width (rhoCurrent/rho0) bwBarrierFac bwBarrierFac;
	 BW = Sqrt[M0 GammaM] / (s0 - sCurrent - I M0 GammaM);
         BW
         ];

(* sinplified barrier function based on s *)
BarrierSimple[s_, sthresh_, s0_, orbMom_]:= If[orbMom==0, 1., ( (s-sthresh) / (s+s0) )^orbMom ];

